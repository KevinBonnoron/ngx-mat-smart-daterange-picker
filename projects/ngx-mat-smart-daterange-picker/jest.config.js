require('jest-preset-angular/ngcc-jest-processor');

module.exports = {
  preset: 'jest-preset-angular',
  transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/html-comment',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/no-ng-attributes',
  ],
};
