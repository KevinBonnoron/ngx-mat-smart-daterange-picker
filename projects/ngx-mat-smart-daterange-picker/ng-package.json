{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/ngx-mat-smart-daterange-picker",
  "lib": {
    "entryFile": "src/public-api.ts"
  }
}