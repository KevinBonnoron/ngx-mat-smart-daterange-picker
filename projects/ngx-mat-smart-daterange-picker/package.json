{
  "name": "ngx-mat-smart-daterange-picker",
  "description": "Angular Material custom daterange picker",
  "version": "0.0.1",
  "publishConfig": {},
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/KevinBonnoron/ngx-mat-smart-daterange-picker"
  },
  "peerDependencies": {
    "@angular/cdk": "^13.0.0",
    "@angular/common": "^13.0.0",
    "@angular/core": "^13.0.0",
    "@angular/material": "^13.0.0",
    "ngx-mat-popover": "0.0.7"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  }
}