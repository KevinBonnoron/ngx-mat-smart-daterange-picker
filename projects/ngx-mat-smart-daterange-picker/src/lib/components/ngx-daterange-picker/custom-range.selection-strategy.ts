import { Injectable } from '@angular/core';
import { DateAdapter } from '@angular/material/core';
import { DateRange, MatDateRangeSelectionStrategy } from '@angular/material/datepicker';

@Injectable()
export class CustomRangeSelectionStrategy<D> implements MatDateRangeSelectionStrategy<D> {
  private _maxRange = 0;

  constructor(private readonly dateAdapter: DateAdapter<D>) {}

  selectionFinished(date: D, currentRange: DateRange<D>) {
    let { start, end } = currentRange;
    if (start === null || (start && end)) {
      start = date;
      end = null;
    } else if (end === null && date !== undefined) {
      const maxDate = this.dateAdapter.addCalendarDays(start, this._maxRange);
      const minDate = this.dateAdapter.addCalendarDays(start, -this._maxRange);
      if (date < start && (this._maxRange === 0 || date >= minDate)) {
        end = start;
        start = date;
      } else if (date >= start && (this._maxRange === 0 || date <= maxDate)) {
        end = date;
      } else {
        start = date;
      }
    }

    return new DateRange<D>(start, end);
  }

  createPreview(activeDate: D | null, currentRange: DateRange<D>): DateRange<D> {
    if (currentRange.start && !currentRange.end) {
      const maxDate = this.dateAdapter.addCalendarDays(currentRange.start, this._maxRange);
      const minDate = this.dateAdapter.addCalendarDays(currentRange.start, -this._maxRange);
      let rangeStart = null;
      let rangeEnd = null;

      if (activeDate !== null) {
        if (activeDate < currentRange.start) {
          rangeStart = this._maxRange !== 0 && activeDate < minDate ? minDate : activeDate;
          rangeEnd = currentRange.start;
        } else {
          rangeStart = currentRange.start;
          rangeEnd = this._maxRange !== 0 && activeDate > maxDate ? maxDate : activeDate;
        }
      }

      return new DateRange(rangeStart, rangeEnd);
    }

    return new DateRange<D>(null, null);
  }

  set maxRange(maxRange: number) {
    this._maxRange = maxRange;
  }
}
