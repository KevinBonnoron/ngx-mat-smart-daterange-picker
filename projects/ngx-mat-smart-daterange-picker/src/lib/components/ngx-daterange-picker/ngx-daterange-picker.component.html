<form [formGroup]="formGroup">
  <mat-form-field style="width: 100%">
    <mat-date-range-input [min]="min" [max]="max" [rangePicker]="picker" style="display: none">
      <input matStartDate formControlName="from" dateFormat="yyyy-MM-dd" (dateChange)="onFromDateChange()" />
      <input matEndDate formControlName="to" dateFormat="yyyy-MM-dd" (dateChange)="onToDateChange($event)" />
    </mat-date-range-input>

    <ngx-mat-date-range-picker #picker [opened]="true"></ngx-mat-date-range-picker>
  </mat-form-field>
</form>