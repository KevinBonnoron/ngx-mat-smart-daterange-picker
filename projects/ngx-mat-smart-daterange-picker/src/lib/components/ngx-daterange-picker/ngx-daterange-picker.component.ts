import { Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges, inject } from '@angular/core';
import { FormBuilder, FormGroup, UntypedFormControl, Validators } from '@angular/forms';
import {
  MAT_DATE_RANGE_SELECTION_STRATEGY,
  MatDateSelectionModel,
  MatDatepickerInputEvent,
  MatRangeDateSelectionModel
} from '@angular/material/datepicker';
import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
import { debounceTime, filter } from 'rxjs/operators';
import { DateRange } from '../../types';
import { isDateRangeEquals } from '../../utils/date-range.utils';
import { CustomRangeSelectionStrategy } from './custom-range.selection-strategy';

@Component({
  selector: 'ngx-daterange-picker',
  templateUrl: './ngx-daterange-picker.component.html',
  providers: [
    {
      provide: MAT_DATE_RANGE_SELECTION_STRATEGY,
      useClass: CustomRangeSelectionStrategy
    },
    {
      provide: MatDateSelectionModel,
      useClass: MatRangeDateSelectionModel
    }
  ]
})
@UntilDestroy()
export class NgxDateRangePickerComponent implements OnInit, OnChanges {
  private readonly formBuilder = inject(FormBuilder);
  private readonly rangeSelectionStrategy = inject(MAT_DATE_RANGE_SELECTION_STRATEGY);

  @Input()
  value!: DateRange;

  @Input()
  min?: Date;

  @Input()
  max?: Date;

  @Input()
  maxRange = 0;

  @Output()
  readonly valueChange = new EventEmitter<DateRange>();

  readonly formGroup: FormGroup = this.formBuilder.group({
    from: [null, Validators.required],
    to: [null, Validators.required]
  });

  ngOnInit(): void {
    this.formGroup.valueChanges
      .pipe(
        debounceTime(50),
        filter(() => this.formGroup.valid),
        filter(() => !isDateRangeEquals(this.value, this.formGroup.value)),
        untilDestroyed(this)
      )
      .subscribe(dateRange => this.valueChange.emit(dateRange));
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes['value']) {
      if (!isDateRangeEquals(this.value, this.formGroup.value)) {
        this.formGroup.patchValue(this.value, { emitEvent: false });
      }
    }

    if (changes['maxRange']) {
      (this.rangeSelectionStrategy as CustomRangeSelectionStrategy<any>).maxRange = this.maxRange;
    }
  }

  onFromDateChange(): void {
    this.to.patchValue(undefined, { emitEvent: false });
  }

  onToDateChange(date: MatDatepickerInputEvent<any>): void {
    if (date.value !== null) {
      this.to.patchValue(new Date(date.value.setHours(23, 59, 59)), { emitEvent: false });
    } else {
      this.to.patchValue(null, { emitEvent: false });
    }
  }

  get from(): UntypedFormControl {
    return this.formGroup.get('from') as UntypedFormControl;
  }

  get to(): UntypedFormControl {
    return this.formGroup.get('to') as UntypedFormControl;
  }
}
