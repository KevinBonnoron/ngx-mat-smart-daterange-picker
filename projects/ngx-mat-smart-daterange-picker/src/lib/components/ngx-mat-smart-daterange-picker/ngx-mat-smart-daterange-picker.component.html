<div>
  <ng-template #popoverContent>
    <div fxFlex="row" style="width: 1100px">
      <div fxFlex="20" fxFlex="1 1 80%">
        <ngx-daterange-picker [value]="value" [min]="min" [max]="max" [maxRange]="maxRange"
          (valueChange)="dateRangePickerChange($event)">
        </ngx-daterange-picker>
      </div>
      <div fxFlex="30" fxFlexOffset="5" fxFlex="column">
        <ngx-absolute-timerange-picker [value]="value" [min]="min" [max]="max"
          (valueChanged)="absoluteTimeRangeChange($event)">
        </ngx-absolute-timerange-picker>
        <ngx-recently-used-time-range [dateRanges]="recentlyUsedTimeRanges"
          (selectionChange)="recentlyUsedTimeRangeChange($event)"></ngx-recently-used-time-range>
      </div>
      <div fxFlex="1 0 auto" fxFlexOffset="5" class="left-separator">
        <ngx-relative-timerange-picker [relativeTimeRanges]="relativeTimeRanges"
          (selectionChange)="relativeTimeRangeChange($event)">
        </ngx-relative-timerange-picker>
      </div>
    </div>
  </ng-template>

  <ngx-mat-popover [popoverContent]="popoverContent" [xPosition]="'after'" [yPosition]="'below'">
    <button mat-raised-button mat-button [disabled]="disabled">
      <mat-icon>schedule</mat-icon>&nbsp;{{ timeRangeLabel }}
    </button>
  </ngx-mat-popover>
</div>