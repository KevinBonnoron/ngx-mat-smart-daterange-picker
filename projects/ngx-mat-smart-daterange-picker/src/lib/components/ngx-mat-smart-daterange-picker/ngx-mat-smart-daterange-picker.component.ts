import { coerceBooleanProperty } from '@angular/cdk/coercion';
import {
  Component,
  ElementRef,
  HostBinding,
  Inject,
  Input,
  LOCALE_ID,
  OnDestroy,
  OnInit,
  Optional,
  Self,
  ViewChild
} from '@angular/core';
import { ControlValueAccessor, NgControl } from '@angular/forms';
import { MatFormFieldControl } from '@angular/material/form-field';
import { NgxMatPopoverComponent } from 'ngx-mat-popover';
import { Subject } from 'rxjs';
import { DateRange, Queue, RelativeTimeRange, UniqueQueue } from '../../types';
import { formatDateRange, isDateRangeEquals } from '../../utils';

@Component({
  selector: 'ngx-mat-smart-daterange-picker',
  templateUrl: './ngx-mat-smart-daterange-picker.component.html',
  providers: [
    {
      provide: MatFormFieldControl,
      useExisting: NgxMatSmartDateRangePickerComponent
    }
  ],
  exportAs: 'ngxSmartDateRangePicker'
})
export class NgxMatSmartDateRangePickerComponent
  implements OnInit, OnDestroy, MatFormFieldControl<DateRange>, ControlValueAccessor
{
  static nextId = 0;

  @ViewChild(NgxMatPopoverComponent, { static: true })
  readonly ngxMatPopover!: NgxMatPopoverComponent;

  @Input()
  min?: Date;

  @Input()
  max?: Date;

  @Input()
  maxRange = 0;

  @Input()
  recentlyUsedTimeRangeCount = 5;

  @Input()
  relativeTimeRanges: RelativeTimeRange[] | undefined = undefined;

  @HostBinding()
  id = `mat-smart-datetime-picker-${NgxMatSmartDateRangePickerComponent.nextId++}`;
  stateChanges = new Subject<void>();
  private _value!: DateRange;
  private _placeholder: string = '';
  private _focused: boolean = false;
  private _required: boolean = false;
  private _disabled = false;
  private onChange: (_: any) => void = () => {};
  private onTouched = () => {};
  controlType = 'mat-smart-datetime-picker';
  autofilled?: boolean | undefined;
  userAriaDescribedBy?: string | undefined;
  recentlyUsedTimeRanges!: Queue<DateRange>;

  constructor(
    @Optional() @Self() public readonly ngControl: NgControl,
    private readonly _elementRef: ElementRef,
    @Inject(LOCALE_ID) private readonly locale: string
  ) {
    if (this.ngControl != null) {
      // Setting the value accessor directly (instead of using
      // the providers) to avoid running into a circular import.
      this.ngControl.valueAccessor = this;
    }
  }

  ngOnInit(): void {
    this.recentlyUsedTimeRanges = new UniqueQueue<DateRange>(this.recentlyUsedTimeRangeCount, isDateRangeEquals);
    if (this.value) {
      this.recentlyUsedTimeRanges.push(this.value);
    }
  }

  ngOnDestroy(): void {
    this.stateChanges.complete();
  }

  dateRangePickerChange(dateRange: DateRange): void {
    this.value = dateRange;
  }

  absoluteTimeRangeChange(dateRange: DateRange): void {
    this.ngxMatPopover.close();
    this.value = dateRange;
  }

  recentlyUsedTimeRangeChange(dateRange: DateRange): void {
    this.ngxMatPopover.close();
    this.value = dateRange;
  }

  relativeTimeRangeChange(dateRange: DateRange): void {
    this.ngxMatPopover.close();
    this.value = dateRange;
  }

  writeValue(value: DateRange): void {
    this._value = value;
  }

  registerOnChange(onChange: any): void {
    this.onChange = onChange;
  }

  registerOnTouched(onTouched: any): void {
    this.onTouched = onTouched;
  }

  setDescribedByIds(ids: string[]): void {
    const controlElement = this._elementRef.nativeElement.querySelector('button')!;
    controlElement.setAttribute('aria-describedby', ids.join(' '));
  }

  onContainerClick(event: MouseEvent): void {
    // Nothing to do
  }

  onFocusIn(event: FocusEvent) {
    if (!this._focused) {
      this._focused = true;
      this.stateChanges.next();
    }
  }

  onFocusOut(event: FocusEvent) {
    if (!this._elementRef.nativeElement.contains(event.relatedTarget as Element)) {
      this._focused = false;
      this.onTouched();
      this.stateChanges.next();
    }
  }

  get timeRangeLabel() {
    if (this.value) {
      return formatDateRange(this.value, this.locale);
    }

    return '';
  }

  @Input()
  get value(): DateRange {
    return this._value;
  }

  set value(value: DateRange) {
    if (value?.from && value?.to && JSON.stringify(value) !== JSON.stringify(this._value)) {
      this.recentlyUsedTimeRanges?.push(value);

      this._value = value;
      this.stateChanges.next();
      this.onChange(value);
    }
  }

  @Input()
  get placeholder() {
    return this._placeholder;
  }

  set placeholder(placeholder) {
    this._placeholder = placeholder;
    this.stateChanges.next();
  }

  get empty(): boolean {
    return this.value === null;
  }

  @HostBinding('class.floating')
  get shouldLabelFloat(): boolean {
    return this._focused || !this.empty;
  }

  @Input()
  get required() {
    return this._required;
  }

  set required(req) {
    this._required = coerceBooleanProperty(req);
    this.stateChanges.next();
  }

  @Input()
  get disabled(): boolean {
    return this._disabled;
  }

  set disabled(value: boolean) {
    this._disabled = coerceBooleanProperty(value);
    // TODO
    // this._disabled ? this.parts.disable() : this.parts.enable();
    this.stateChanges.next();
  }

  get errorState(): boolean {
    return this.ngControl?.invalid ?? false;
  }

  get focused(): boolean {
    return this._focused;
  }
}
