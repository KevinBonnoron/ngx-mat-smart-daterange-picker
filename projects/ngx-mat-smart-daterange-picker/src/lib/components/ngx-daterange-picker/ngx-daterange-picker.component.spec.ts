import { ReactiveFormsModule } from '@angular/forms';
import { DateAdapter, NativeDateAdapter } from '@angular/material/core';
import { MockBuilder, MockRenderFactory, ngMocks } from 'ng-mocks';
import { NgxSmartDateRangePickerModule } from '../../ngx-mat-smart-daterange-picker.module';
import { NgxDateRangePickerComponent } from './ngx-daterange-picker.component';

describe('NgxDateRangePickerComponent', () => {
  const factory = MockRenderFactory(NgxDateRangePickerComponent);

  ngMocks.faster();

  beforeAll(() =>
    MockBuilder(NgxDateRangePickerComponent, NgxSmartDateRangePickerModule)
      .keep(ReactiveFormsModule)
      .mock(DateAdapter, NativeDateAdapter)
  );
  beforeAll(() => factory.configureTestBed());

  it('should create', () => {
    const fixture = factory();
    expect(fixture).toBeDefined();
  });

  it('should have the expected render', () => {
    const fixture = factory();
    expect(fixture).toMatchSnapshot();
  });
});
