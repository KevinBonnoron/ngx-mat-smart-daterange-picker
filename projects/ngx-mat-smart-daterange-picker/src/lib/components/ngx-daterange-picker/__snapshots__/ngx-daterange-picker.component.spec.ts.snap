// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgxDateRangePickerComponent should have the expected render 1`] = `
<mock-render
  __ngContext__={[Function LRootView]}
>
  <ngx-daterange-picker>
    <form
      class="ng-untouched ng-pristine ng-invalid"
      novalidate=""
    >
      <mat-form-field
        style="width: 100%;"
      >
        <mat-date-range-input
          style="display: none;"
        >
          <input
            class="ng-untouched ng-pristine ng-invalid"
            dateformat="yyyy-MM-dd"
            formcontrolname="from"
            matstartdate=""
          />
          <input
            class="ng-untouched ng-pristine ng-invalid"
            dateformat="yyyy-MM-dd"
            formcontrolname="to"
            matenddate=""
          />
        </mat-date-range-input>
        <ngx-mat-date-range-picker />
      </mat-form-field>
    </form>
  </ngx-daterange-picker>
</mock-render>
`;
